// spfx-library/src/libraries/spfxLibrary/helpers/Constants.ts
/**
 * Constants class
 */
export class Constants {
  public static ApplicationName = "HCL Viva";
  public static AutorityUrl = "https://login.microsoftonline.com/{tenant-id}";
  public static TenantIdPlaceHolder = "{tenant-id}";
  public static APITimeoutMessage = "API Timeout";
  public static DefaultExceptionString = "Error occured in ";
  public static GraphAPI_MyProfile = "https://graph.microsoft.com/v1.0/me";
  public static MyProfileGraphPermission = "user.read";
  public static GraphAPI_MyProfileErrorMessage = "Error my profile :";
  public static UpdateOpportunityMessage = " value getting closed in Next 15 days.";
  public static RightArrowImage = "/SiteAssets/Images/rightArrow.png";
  public static VerticalLineImage = "/SiteAssets/Images/verticalLine.png";
  public static NullOrEmptyData = "No data available";
  public static Param_TravelDetails = "?source=";
  public static Title_TravelDetails = "Travel details";
  public static SyncLabel ="Syncing";
  public static NoPendingActionMessage= "There are no pending actions.";
  public static VivaConnectionJS= "/SiteAssets/CustomTheme/style/VivaConnection.js";
  public static UserProfileProperty= "hcltech-restricteddata";
  public static GuidForCompliantUser = "1cd247c5-67eb-4bb5-911b-8120401a0dd1";
  public static ListIDForExtension = "bdb4bbb3-8aba-45a5-9fa9-7fbeb5aba6fd";
  public static selectFieldsFromList = "Title,HCLTechConfigValue";
  public static AllowedLoginsColumn = "AllowedLogins";
  public static TenantId = "TenantId";
  public static ClientId = "ClientId";
  public static AADRedirectUrl = "AADRedirectUrl";
  public static Scope = "Scope";
  public static OcpApimTrace = "OcpApimTrace";
  public static OcpApimSubscriptionKey = "OcpApimSubscriptionKey";
  public static ComplianceAPIURL = "ComplianceAPIURL";
  public static APImethod = "APImethod";
  public static APITimeoutLimit = "APITimeoutLimit";
  public static PageToSkip = "PageToSkip";
  public static RedirectURL = "RedirectURL";
  public static CacheStorageExpiry = "CacheStorageExpiry";
  public static GuidForNonCompliantUser = "a5d95a05-e44e-4c68-aeef-9b279f32882d";
  public static HomePage="Home.aspx";
  public static SPOListForConfiguration = "HCLTechExtensionConfig";
  public static NonCompliant_WarnImage = "/SiteAssets/Images/warn.png";
  public static LeaveBalanceCacheKey = "HCL_LeaveBalance_Info";
  public static PendingApprovalCacheKey = "HCL_PendingApproval_Info";
  public static MyContactCacheKey = "HCL_MyContact_Info";
  public static PendingAction_TimesheetCompliance_CacheKey = "HCL_TimesheetCompliance_Info";
  public static PendingAction_UpdateOpportunity_CacheKey = "HCL_UpdateOpportunity_Info";
  public static PendingAction_MyPerformance_CacheKey = "HCL_MyPerformance_Info";
  public static PendingAction_TravelDetails_CacheKey = "HCL_TravelDetails_Info";
  public static EmployeeCode_CacheKey = "HCL_EmployeeCode_Info";
  public static MyTraining_CacheKey = "HCL_MyTraining_Info";
  public static ComplianceAPI_CacheKey = "HCL_ComplianceAPI_Info";
  public static UserProfile_CacheKey = "HCL_UserProfile_Info";
  public static UserProfileFlag_CacheKey = "HCL_UserProfileFlag_Info";
  public static UpcomingHolidaysCacheKey = "HCL_UpcominmgHolidays_Info";
  public static DashboardStatus_SSDAPI_CacheKey = "HCL_DBStatusSSD_Info";
  public static DashboardStatus_LMSAPI_CacheKey = "HCL_DBStatusLMS_Info";
  public static NullValue = "NullValue";
  public static IsCompliantFlag_CacheKey = "HCL_IsCompliantFlag_Info";
  //Compliance cache expiration time
  public static Compliance_ExpirationTime_CacheKey = "1";


  /**
   * Notification control of constants
   */
  public static NotificationControl = {
    ModifiedBy: "Modified",
    IsActive: "HCLTechItemIsActive",
    StartDate: "HCLTechItemStartDate",
    EndDate: "HCLTechItemEndDate",
    DefaultAnimationDuration: "6s",
  };


  /**
   * Dashboard holidays of constants
   */
  public static DashboardHolidays = {
    DefaultExceptionString: "Error occured in Dashboard Holidays",
    APITimeoutMessage: "API Timeout",
    SyncLabel: "Syncing",
    NoUpcomingHolidayMessage: "There are no upcoming holidays.",
    SeeAllLabel: "See all",
    MoreLabel: "More",
    OpenLMSLabel: "Open LMS",
    ApplyRHLabel:"Apply RH"
  };


  /**
   * Dashboard status of constants
   */
  public static DashboardStatus = {
    DefaultExceptionString: "Error occured in DashboardStatus ",
  };


  /**
   * Leave balance of constants
   */
  public static LeaveBalance = {
    DefaultExceptionString: "Error occured in ",
    APITimeoutMessage: "API Timeout",
  };
  public static MyTrainings = {
    DefaultExceptionString: "Error occured in ",
    APITimeoutMessage: "API Timeout",
  };


  /**
   * Determines whether null or undefined or empty is
   * @param value 
   * @returns true if null or undefined or empty 
   */
  public static IsNullOrUndefinedOrEmpty(value: any): boolean {
    return (value == null || value === "" || value === undefined);
  }

}
